astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -685,16 +686,27 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
         if clickhouse_manifest.version != ClickHouseBackupVersion.V1:
             return
         tables_by_uuid = {table.uuid: table for table in clickhouse_manifest.tables}
+
+        async def process_part(part_desc: tuple[str, bytes], client: ClickHouseClient) -> None:
+            table_identifier, part_name = part_desc
+            await execute_with_timeout(
+                client,
+                self.attach_timeout,
+                f"ALTER TABLE {table_identifier} ATTACH PART {escape_sql_string(part_name)}".encode(),
+            )
+
         await gather_limited(
             self.max_concurrent_attach,
             [
-                execute_with_timeout(
-                    client,
-                    self.attach_timeout,
-                    f"ALTER TABLE {table_identifier} ATTACH PART {escape_sql_string(part_name)}".encode(),
+                process_part(part_desc, client)
+                for part_descs in zip_longest(
+                    *[
+                        list_parts_to_attach(snapshot_result, self.disks, tables_by_uuid)
+                        for snapshot_result in backup_manifest.snapshot_results
+                    ]
```

Comment:
This change is good but I think it becomes hard to follow with the `zip_longest` and it does not fully reach the apparent goal of having at most `max_concurrent_attach/node_count` concurrent queries on each node.

Maybe something like that instead :
```python
        # Do all nodes at the same time
        await asyncio.gather(
            *[
                # But on each node, limit the number of concurrent attach
                gather_limited(
                    self.max_concurrent_attach_per_node,
                    [
                        execute_with_timeout(
                            client,
                            self.attach_timeout,
                            f"ALTER TABLE {table_identifier} ATTACH PART {escape_sql_string(part_name)}".encode(),
                        )
                        for table_identifier, part_name in
                        list_parts_to_attach(snapshot_result, self.disk_paths, tables_by_uuid)
                    ]
                )
                for client, snapshot_result in zip(self.clients, backup_manifest.snapshot_results)
            ]
        )
```

(In the context of Astacus, it's ok to create a new setting and deprecate/ignore the old one)

Reply:
Thanks! It looks like I was a bit too focused on transposing the loops and missed the whole point. Fixed.